### ISSUE TITLE
Custom loss with dependent samples

### ISSUE THREAD

**Issue by voilasept** [microsoft/LightGBM#6145]
All the examples I've seen are custom loss function where samples are independent, such as MSE. Sometimes the custom loss has dependency between samples. 
The dependency is in batches. Say, we have 1 year of data. For each day, the samples have dependency, but not across different days.

The loss for each day looks like this:
Loss(a day) = (y1-y0)^2 + (y2-y1)^2 + ...
Total loss:
Loss = Loss(day 0) + Loss(day 1) + ... 

Is this implementable under current python API? Thank you!

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

Nothing about the way the Python package supports custom objective functions assumes independence between observations in the training data, or require you to implement pointwise loss.

As described at https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.train.html#lightgbm.train, using `lightgbm.train()`, it's possible to provide a Python function which accepts the predictions as of the current iteration **and the full training dataset**, and which returns the gradient and hessian for the loss.

Since that function can access the entire training `Dataset`, there's nothing stopping you from implementing a loss function which considers multiple observations in the training data. It's just up to you to figure out how to do that in a way that the gradients are informative for the purpose of training the model.

Here's an example of a custom loss function using LightGBM's Python package:

https://github.com/microsoft/LightGBM/blob/8ed371cee49cf86740b25dd9a4b985a75c9f2dba/examples/python-guide/advanced_example.py#L139-L147
